### PR TITLE
General additions 2

### PR DESCRIPTION
**Pull request by ahuth**
Potential replacement of #13.

The master branch has some weird stuff on it, preventing the `general_additions` branch from merging cleanly. Here, we've merged master into `general_additions` with `strategy=ours`.

### CODE REVIEW

**Comment by ahuth** [comment]
@shirish-pampoorickal Can you take a look at this please?

We'll also QA this.

File: lib/config/specification.yml [hunk 1]
```diff
@@ -1,4 +1,70 @@
 ---
+account_invitations:
+  attributes:
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: Id missing for account_invitations, additional_items & account_memberships

File: lib/config/specification.yml [hunk 2]
```diff
@@ -1,4 +1,70 @@
...
+    inviter:
+      foreign_key: inviter_id
+      collection: users
+    invitee:
+      foreign_key: invitee_id
+      collection: users
+    default_role:
+      foreign_key: default_role_id
+      collection: roles
+additional_items:
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: Add Line break 👮

File: lib/config/specification.yml [hunk 3]
```diff
@@ -142,6 +400,64 @@ invoices:
       foreign_key: recipient_id
       collection: users
 
+organizations:
+  attributes:
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: id missing?

File: lib/config/specification.yml [hunk 4]
```diff
@@ -142,6 +400,64 @@ invoices:
...
+    - ancestor_ids
+  create_attributes:
+    - name
+    - parent_id
+  update_attributes:
+    - name
+    - parent_id
+
+organization_memberships:
+  attributes:
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: id missing

File: lib/config/specification.yml [hunk 5]
```diff
@@ -213,7 +529,151 @@ posts:
...
+rate_card_sets:
+  associations:
+    rate_cards:
+      foriegn_key: rate_card_ids
+      collection: rate_cards
+    rate_card_set_versions:
+      foreign_key: rate_card_set_version_ids
+      collection: rate_card_set_versions
+  attributes:
+    - title
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: id missing

File: lib/config/specification.yml [hunk 6]
```diff
@@ -213,7 +529,151 @@ posts:
...
+rate_card_set_versions:
+  associations:
+    rate_card_set:
+      foreign_key: rate_card_set_id
+      collection: rate_card_sets
+    rate_card_versions:
+      foreign_key: rate_card_version_ids
+      collection: rate_card_versions
+  attributes:
+    - active
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: id missing

File: lib/config/specification.yml [hunk 7]
```diff
@@ -213,7 +529,151 @@ posts:
...
+      foreign_key: rate_card_id
+      collection: rate_cards
+    rate_card_set_version:
+      foreign_key: rate_card_set_version_id
+      collection: rate_card_set_versions
+    rate_card_roles:
+      foreign_key: rate_card_role_ids
+      collection: rate_card_roles
+  attributes:
+    - default_rate
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: id missing

File: lib/config/specification.yml [hunk 8]
```diff
@@ -213,7 +529,151 @@ posts:
...
+  create_attributes:
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: Shouldn't this have default_rate as an attribute?

File: lib/config/specification.yml [hunk 9]
```diff
@@ -213,7 +529,151 @@ posts:
...
+
+rate_card_roles:
+  associations:
+    role:
+      foreign_key: role_id
+      collection: roles
+    rate_card_version:
+      foreign_key: rate_card_version_id
+      collection: rate_card_versions
+  attributes:
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: id missing

File: lib/config/specification.yml [hunk 10]
```diff
@@ -213,7 +529,151 @@ posts:
...
+  create_attributes:
+    - rate_card_version_id
+    - role_id
+    - rate
+  update_attributes:
+    - rate
+
+resolutions:
+  attributes:
+    - type
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: id missing

File: lib/config/specification.yml [hunk 11]
```diff
@@ -213,7 +529,151 @@ posts:
...
+    - created_at
+    - updated_at
+    - description
+    - target_type
+    - target_id
+    - user_id
+
 roles: # TODO(SZ): find it out
+  attributes:
+    - name
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: id missing

File: lib/config/specification.yml [hunk 12]
```diff
@@ -397,6 +867,48 @@ time_entries:
...
+      time_entries:
+        foreign_key: time_entry_ids
+        collection: time_entries
+      resolutions:
+        foreign_key: resolution_ids
+        collection: resolutions
+
 users:
   attributes:
     - user_id       # the internal ID of the user
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: id missing

File: lib/config/specification.yml [hunk 13]
```diff
@@ -454,40 +973,115 @@ workspaces:
...
+    - consultant_role_name
+    - posts_require_privacy_decision
+    - stories_are_fixed_fee_by_default
+    - currency
+    - require_expense_approvals
+    - require_time_approvals
+
+workspace_invoice_preferences:
+  attributes:
+    - workspace_id
```
Comment (shirish-pampoorickal) [inline lib/config/specification.yml]: id missing

File: lib/mavenlink/concerns/custom_fieldable.rb [hunk 14]
```diff
@@ -0,0 +1,10 @@
+module Mavenlink
+  module Concerns
+    module CustomFieldable
+
```
Comment (shirish-pampoorickal) [inline lib/mavenlink/concerns/custom_fieldable.rb]: Unnecessary line break 👮

File: lib/mavenlink/model.rb [hunk 15]
```diff
@@ -232,7 +229,7 @@ def attributes_for_update
 
     # @return [Mavenlink::Request]
     def request
-      scoped_im.only(id)
+      collection_scope.only(id)
```
Comment (shirish-pampoorickal) [inline lib/mavenlink/model.rb]: We should remove `scoped_im` method if it is not being used anywhere

File: spec/lib/mavenlink/client_spec.rb [hunk 16]
```diff
@@ -13,6 +13,55 @@
...
+      client = described_class.new(oauth_token: '12345', endpoint: endpoint)
+
+      expect(client.send(:endpoint)).to eq endpoint
+    end
+  end
+
+  describe 'association calls' do
+    subject(:client) { described_class.new(oauth_token: '12345') }
+    let(:record) { client.workspaces.find(7) }
+
```
Comment (shirish-pampoorickal) [inline spec/lib/mavenlink/client_spec.rb]: Unnecessary line break 👮

File: spec/lib/mavenlink/model_spec.rb [hunk 17]
```diff
@@ -182,6 +182,18 @@ def self.name
...
 
+  describe '#client' do
+    specify 'default client' do
+      expect(model.new.client).to be_a Mavenlink::Client
+    end
+
+    context 'custom client set' do
+      let(:client) { Mavenlink::Client.new(oauth_token: 'new one') }
+      subject { described_class.new({test: 'set'}, nil, client) }
+      its(:client) { should eq(client) }
```
Comment (shirish-pampoorickal) [inline spec/lib/mavenlink/model_spec.rb]: line break please 👮

File: spec/lib/mavenlink/response_spec.rb [hunk 18]
```diff
@@ -32,5 +32,18 @@
...
+
+  describe '#client' do
+    specify 'default client' do
+      expect(response.client).to be_a Mavenlink::Client
+    end
+
+    context 'custom client set' do
+      let(:client) { Mavenlink::Client.new(oauth_token: 'new one') }
+      subject(:response) { described_class.new(response_data, client) }
+      its(:client) { should eq(client) }
```
Comment (shirish-pampoorickal) [inline spec/lib/mavenlink/response_spec.rb]: Line break please 👮

File: lib/config/specification.yml (hunk 8 above)
Comment (ahuth) [inline lib/config/specification.yml]: Seems that way!

File: spec/lib/mavenlink/response_spec.rb (hunk 18 above)
Comment (ahuth) [inline spec/lib/mavenlink/response_spec.rb]: I think this one is right as well

File: spec/lib/mavenlink/model_spec.rb (hunk 17 above)
Comment (shirish-pampoorickal) [inline spec/lib/mavenlink/model_spec.rb]: Normal specs look like:
```
context "...." do
  let(:blah) { "blah" }
  subject { blah }
---> Line break
  it "does something"
end
```

Based on this pattern I suggested, we do:
```
context 'custom client set' do
  let(:client) { Mavenlink::Client.new(oauth_token: 'new one') }
  subject(:response) { described_class.new(response_data, client) }
  
  its(:client) { should eq(client) }
end
```

NOTE: I am totally fine, if you guys think there should be no line break

Comment (ahuth) [inline spec/lib/mavenlink/model_spec.rb]: Oh, I thought you were asking for one after the lets.

I'll add the line break. That makes more sense

File: lib/mavenlink/model.rb (hunk 15 above)
Comment (ahuth) [inline lib/mavenlink/model.rb]: Not sure what it's for, but it is being tested

**Comment by ahuth** [comment]
Looks like the tests haven't been ran in a long time. In order to run them, I had to remove the Guard gem, as it's not compatible with Ruby 2.2.3.

Once those were running, there were a bunch of failures. They've now been fixed up, and mergin...

**Pull request merged**